h3/src/connection.rs

```diff
@@ -460,8 +433,43 @@ where
         Poll::Ready(res)
     }
 
-    pub fn start_stream(&mut self, id: StreamId) {
-        self.last_accepted_stream = Some(id);
+    pub(crate) fn process_goaway<T>(
+        &mut self,
+        recv_closing: &mut Option<T>,
+        id: VarInt,
+    ) -> Result<(), Error>
+    where
+        T: From<VarInt> + Copy,
+        VarInt: From<T>,
+    {
+        {
+            //= https://www.rfc-editor.org/rfc/rfc9114#section-5.2
+            //# An endpoint MAY send multiple GOAWAY frames indicating different
+            //# identifiers, but the identifier in each frame MUST NOT be greater
+            //# than the identifier in any previous frame, since clients might
+            //# already have retried unprocessed requests on another HTTP connection.
+
+            //= https://www.rfc-editor.org/rfc/rfc9114#section-5.2
+            //# Like the server,
+            //# the client MAY send subsequent GOAWAY frames so long as the specified
+            //# push ID is no greater than any previously sent value.
+            if let Some(prev_id) = recv_closing.map(VarInt::from) {
+                if prev_id < id {
+                    //= https://www.rfc-editor.org/rfc/rfc9114#section-5.2
+                    //# Receiving a GOAWAY containing a larger identifier than previously
+                    //# received MUST be treated as a connection error of type H3_ID_ERROR.
+                    return Err(self.close(
+                        Code::H3_ID_ERROR,
+                        format!(
+                            "received a GoAway({}) greater than the former one ({})",
+                            id, prev_id
+                        ),
+                    ));
+                }
+            }
+            *recv_closing = Some(id.into());
+            Ok(())
```

Comment:
Thanks for this awesome PR.

I just got one question, do we need a additional `self.shared.write("connection goaway write").closing = true;` here?

Reply:
No we don't. This is done on frame receipt on connection level in poll_control (same file L378). That prevented duplication before I merged the client and server side code in the generic `process_goaway`. Maybe I should move it to the place you suggested to keep those things closer together like it is done on shutdown. Let me know what you think. I'm slightly in favor of moving it.

Reply:
Oh I missed that. 
But I am also thinking that it could be slightly better to move it.

Reply:
done